src/abstractdataframe/join.jl

```diff
@@ -116,34 +118,24 @@ Base.IndexStyle(::Type{<:OnCol}) = Base.IndexLinear()
 
 @inline function Base.getindex(oc::OnCol, i::Int)
     @boundscheck checkbounds(oc, i)
-    return OnColRow(i, oc.cols)
+    return OnColRow(i, oc.cols, oc.h)
 end
 
-# TODO: rewrite hash, isequal and isless to use @generated
-# or some other approach that would keep them efficient and avoid code duplication
-
-@inline function Base.hash(ocr1::OnColRow{<:NTuple{2, AbstractVector}}, h::UInt)
-    r1 = ocr1.row
-    c11, c12 = ocr1.cols
-    return @inbounds hash(c11[r1], hash((c12[r1],), h))
-end
-
-@inline function Base.hash(ocr1::OnColRow{<:NTuple{3,AbstractVector}}, h::UInt)
-    r1 = ocr1.row
-    c11, c12, c13 = ocr1.cols
-    return @inbounds hash(c11[r1], hash(c12[r1], hash((c13[r1],), h)))
-end
+Base.hash(ocr1::OnColRow, h::UInt) = throw(MethodError(hash, (ocr1, h)))
+@inline Base.hash(ocr1::OnColRow) = @inbounds ocr1.h[ocr1.row]
 
-@inline function Base.hash(ocr1::OnColRow{<:NTuple{N,AbstractVector}}, h::UInt) where {N}
-    r1 = ocr1.row
-    cols1 = ocr1.cols
-    @inbounds hv = hash((cols1[end][r1],), h)
-    for i in N-1:-1:1
-        hv = @inbounds hash(cols1[i][r1], hv)
+function _prehash(oc::OnCol)
+    h = oc.h
+    resize!(h, oc.len)
+    fill!(h, Base.tuplehash_seed)
+    for col in reverse(oc.cols)
+        h .= hash.(col, h)
     end
-    return return hv
 end
 
+# TODO: rewrite isequal and isless to use @generated
+# or some other approach that would keep them efficient and avoid code duplication
```

Comment:
These seem to be about as fast as the manually unrolled version (a bit slower for `Float64` but a bit faster for `Int` for some reason):
```julia
function Base.isequal(ocr1::OnColRow{<:NTuple{N,AbstractVector}},
                      ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where N
    return isequal(ntuple(i -> @inbounds(ocr1.cols[i][ocr1.row]), N),
                   ntuple(i -> @inbounds(ocr2.cols[i][ocr2.row]), N))
end

function Base.isless(ocr1::OnColRow{<:NTuple{N,AbstractVector}},
                     ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where N
    return isless(ntuple(i -> @inbounds(ocr1.cols[i][ocr1.row]), N),
                  ntuple(i -> @inbounds(ocr2.cols[i][ocr2.row]), N))
end
```

Reply:
I was checking it and it was slower, but I will check again and report the results here

Reply:
Here are the benchmarks (the conclusion is that my current approach with manual unrolling is never very bad, sometimes tuple is better, but it seems that what you propose is worst):
```
julia> using DataFrames, BenchmarkTools

julia> import DataFrames: OnCol, OnColRow

julia> function fisequal(y)
                 s = 0
                 for x in y, v in y
                     s += isequal(v, x)
                 end
                 return s
              end
fisequal (generic function with 1 method)

julia> function fisless(y)
                 s = 0
                  for x in y, v in y
                     s += isless(v, x)
                 end
                 return s
              end
fisless (generic function with 1 method)

julia> function fisequal2(y)
                 s = 0
                 for x in y, v in y
                     s += isequal2(v, x)
                 end
                 return s
              end
fisequal2 (generic function with 1 method)

julia> function fisless2(y)
                 s = 0
                  for x in y, v in y
                     s += isless2(v, x)
                 end
                 return s
              end
fisless2 (generic function with 1 method)

julia> isequal2(ocr1::OnColRow{<:NTuple{N,AbstractVector}},
                    ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where {N} =
           isequal(ntuple(i -> @inbounds(ocr1.cols[i][ocr1.row]), N),
                   ntuple(i -> @inbounds(ocr2.cols[i][ocr2.row]), N))
isequal2 (generic function with 1 method)

julia> isless2(ocr1::OnColRow{<:NTuple{N,AbstractVector}},
                   ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where {N} =
           isless(ntuple(i -> @inbounds(ocr1.cols[i][ocr1.row]), N),
                  ntuple(i -> @inbounds(ocr2.cols[i][ocr2.row]), N))
isless2 (generic function with 1 method)

julia> x = rand(10^3);

julia> y = rand(10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  1.770 ms (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  1.408 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  1.516 ms (0 allocations: 0 bytes)

julia> @btime fisless($oc);
  5.287 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  5.766 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  4.793 ms (0 allocations: 0 bytes)

julia> x = rand([1.0:10.0; missing], 10^3);

julia> y = rand([1.0:10.0; missing], 10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  25.209 ms (1976048 allocations: 30.15 MiB)

julia> @btime fisequal2($oc);
  182.640 ms (5598000 allocations: 110.17 MiB)

julia> @btime fisequal($t);
  36.490 ms (944739 allocations: 14.42 MiB)

julia> @btime fisless($oc);
  18.895 ms (988426 allocations: 15.08 MiB)

julia> @btime fisless2($oc);
  183.023 ms (5598000 allocations: 110.17 MiB)

julia> @btime fisless($t);
  40.092 ms (998267 allocations: 15.23 MiB)

julia> x = rand(Int, 10^3);

julia> y = rand(Int, 10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  765.115 μs (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  1.756 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  259.885 μs (0 allocations: 0 bytes)

julia> @btime fisless($oc);
  2.988 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  3.059 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  1.928 ms (0 allocations: 0 bytes)

julia> x = rand([1:10; missing], 10^3);

julia> y = rand([1:10; missing], 10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  3.373 ms (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  3.600 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  35.263 ms (947554 allocations: 14.46 MiB)

julia> @btime fisless($oc);
  4.221 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  4.773 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  38.683 ms (998791 allocations: 15.24 MiB)
```

The reason seems to be:
```
julia> @code_warntype isequal2(oc[1], oc[2])
Variables
  #self#::Core.Compiler.Const(isequal2, false)
  ocr1::OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}
  ocr2::OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}
  #33::var"#33#35"{OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}}
  #34::var"#34#36"{OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}}

Body::Bool
1 ─ %1  = Main.:(var"#33#35")::Core.Compiler.Const(var"#33#35", false)
│   %2  = Core.typeof(ocr1)::Core.Compiler.Const(OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}, false)
│   %3  = Core.apply_type(%1, %2)::Core.Compiler.Const(var"#33#35"{OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}}, false)
│         (#33 = %new(%3, ocr1))
│   %5  = #33::var"#33#35"{OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}}
│   %6  = Main.ntuple(%5, $(Expr(:static_parameter, 1)))::Tuple{Union{Missing, Int64},Union{Missing, Int64}}
│   %7  = Main.:(var"#34#36")::Core.Compiler.Const(var"#34#36", false)
│   %8  = Core.typeof(ocr2)::Core.Compiler.Const(OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}, false)
│   %9  = Core.apply_type(%7, %8)::Core.Compiler.Const(var"#34#36"{OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}}, false)
│         (#34 = %new(%9, ocr2))
│   %11 = #34::var"#34#36"{OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}}
│   %12 = Main.ntuple(%11, $(Expr(:static_parameter, 1)))::Tuple{Union{Missing, Int64},Union{Missing, Int64}}
│   %13 = Main.isequal(%6, %12)::Bool
└──       return %13

julia> @code_warntype isequal(oc[1], oc[2])
Variables
  #self#::Core.Compiler.Const(isequal, false)
  ocr1::OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}
  ocr2::OnColRow{Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}}
  @_4::Int64
  @_5::Int64
  val::Bool
  r1::Int64
  c11::Array{Union{Missing, Int64},1}
  c12::Array{Union{Missing, Int64},1}
  r2::Int64
  c21::Array{Union{Missing, Int64},1}
  c22::Array{Union{Missing, Int64},1}
  @_13::Bool

Body::Bool
1 ─       nothing
│         Core.NewvarNode(:(val))
│         (r1 = Base.getproperty(ocr1, :row))
│   %4  = Base.getproperty(ocr1, :cols)::Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}
│   %5  = Base.indexed_iterate(%4, 1)::Core.Compiler.PartialStruct(Tuple{Array{Union{Missing, Int64},1},Int64}, Any[Array{Union{Missing, Int64},1}, Core.Compiler.Const(2, false)])
│         (c11 = Core.getfield(%5, 1))
│         (@_4 = Core.getfield(%5, 2))
│   %8  = Base.indexed_iterate(%4, 2, @_4::Core.Compiler.Const(2, false))::Core.Compiler.PartialStruct(Tuple{Array{Union{Missing, Int64},1},Int64}, Any[Array{Union{Missing, Int64},1}, Core.Compiler.Const(3, false)])
│         (c12 = Core.getfield(%8, 1))
│         (r2 = Base.getproperty(ocr2, :row))
│   %11 = Base.getproperty(ocr2, :cols)::Tuple{Array{Union{Missing, Int64},1},Array{Union{Missing, Int64},1}}
│   %12 = Base.indexed_iterate(%11, 1)::Core.Compiler.PartialStruct(Tuple{Array{Union{Missing, Int64},1},Int64}, Any[Array{Union{Missing, Int64},1}, Core.Compiler.Const(2, false)])
│         (c21 = Core.getfield(%12, 1))
│         (@_5 = Core.getfield(%12, 2))
│   %15 = Base.indexed_iterate(%11, 2, @_5::Core.Compiler.Const(2, false))::Core.Compiler.PartialStruct(Tuple{Array{Union{Missing, Int64},1},Int64}, Any[Array{Union{Missing, Int64},1}, Core.Compiler.Const(3, false)])
│         (c22 = Core.getfield(%15, 1))
│         $(Expr(:inbounds, true))
│   %18 = Base.getindex(c11, r1)::Union{Missing, Int64}
│   %19 = Base.getindex(c21, r2)::Union{Missing, Int64}
│   %20 = DataFrames.isequal(%18, %19)::Bool
└──       goto #3 if not %20
2 ─ %22 = Base.getindex(c12, r1)::Union{Missing, Int64}
│   %23 = Base.getindex(c22, r2)::Union{Missing, Int64}
│         (@_13 = DataFrames.isequal(%22, %23))
└──       goto #4
3 ─       (@_13 = false)
4 ┄       (val = @_13)
│         $(Expr(:inbounds, :pop))
└──       return val
```
And the compiler in case of your approach highlights:
```
Tuple{Union{Missing, Int64},Union{Missing, Int64}}
```
in red, which I assume signals that less optimization is performed for it.

Reply:
Ah that's funny. This is on 1.5, right? I had only tried on a recent main. Looks like things improved a lot! Here's what I get on 1.6.0-beta1. Overall my proposal is a bit slower, except for `Int` where it's slower for `isequal` but faster for `isless`. That difference is suprising.
```julia
julia> using DataFrames, BenchmarkTools

julia> import DataFrames: OnCol, OnColRow

julia> function fisequal(y)
                 s = 0
                 for x in y, v in y
                     s += isequal(v, x)
                 end
                 return s
              end
fisequal (generic function with 1 method)

julia> function fisless(y)
                 s = 0
                  for x in y, v in y
                     s += isless(v, x)
                 end
                 return s
              end
fisless (generic function with 1 method)

julia> function fisequal2(y)
                 s = 0
                 for x in y, v in y
                     s += isequal2(v, x)
                 end
                 return s
              end
fisequal2 (generic function with 1 method)

julia> function fisless2(y)
                 s = 0
                  for x in y, v in y
                     s += isless2(v, x)
                 end
                 return s
              end
fisless2 (generic function with 1 method)

julia> isequal2(ocr1::OnColRow{<:NTuple{N,AbstractVector}},
                    ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where {N} =
           isequal(ntuple(i -> @inbounds(ocr1.cols[i][ocr1.row]), N),
                   ntuple(i -> @inbounds(ocr2.cols[i][ocr2.row]), N))
isequal2 (generic function with 1 method)

julia> isless2(ocr1::OnColRow{<:NTuple{N,AbstractVector}},
                   ocr2::OnColRow{<:NTuple{N,AbstractVector}}) where {N} =
           isless(ntuple(i -> @inbounds(ocr1.cols[i][ocr1.row]), N),
                  ntuple(i -> @inbounds(ocr2.cols[i][ocr2.row]), N))
isless2 (generic function with 1 method)

julia> x = rand(10^3);

julia> y = rand(10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  2.510 ms (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  2.248 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  2.481 ms (0 allocations: 0 bytes)

julia> @btime fisless($oc);
  6.826 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  8.346 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  6.554 ms (0 allocations: 0 bytes)

julia> x = rand([1.0:10.0; missing], 10^3);

julia> y = rand([1.0:10.0; missing], 10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  5.223 ms (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  5.699 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  46.430 ms (946931 allocations: 14.45 MiB)

julia> @btime fisless($oc);
  9.199 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  10.849 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  54.736 ms (997992 allocations: 15.23 MiB)

julia> x = rand(Int, 10^3);

julia> y = rand(Int, 10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  842.428 μs (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  2.017 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  341.707 μs (0 allocations: 0 bytes)

julia> @btime fisless($oc);
  3.552 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  1.482 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  512.004 μs (0 allocations: 0 bytes)

julia> x = rand([1:10; missing], 10^3);

julia> y = rand([1:10; missing], 10^3);

julia> oc = OnCol(x, y);

julia> t = tuple.(x,y);

julia> @btime fisequal($oc);
  4.140 ms (0 allocations: 0 bytes)

julia> @btime fisequal2($oc);
  5.587 ms (0 allocations: 0 bytes)

julia> @btime fisequal($t);
  44.686 ms (943288 allocations: 14.39 MiB)

julia> @btime fisless($oc);
  6.513 ms (0 allocations: 0 bytes)

julia> @btime fisless2($oc);
  7.854 ms (0 allocations: 0 bytes)

julia> @btime fisless($t);
  52.752 ms (999349 allocations: 15.25 MiB)
```

Reply:
I have made some tests on nightly and they were similar.
What I left is special casing for 2 columns and using your implementation for more than 2.

Reply:
Sounds like a good compromise. :-)